keyboards/handwired/battleship_gamepad/keymaps/default/keymap.c

```diff
@@ -0,0 +1,42 @@
+/* Copyright 2021 Andrew Braini
+  * 
+  * This program is free software: you can redistribute it and/or modify 
+  * it under the terms of the GNU General Public License as published by 
+  * the Free Software Foundation, either version 2 of the License, or 
+  * (at your option) any later version. 
+  * 
+  * This program is distributed in the hope that it will be useful, 
+  * but WITHOUT ANY WARRANTY; without even the implied warranty of 
+  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
+  * GNU General Public License for more details. 
+  * 
+  * You should have received a copy of the GNU General Public License 
+  * along with this program.  If not, see <http://www.gnu.org/licenses/>. 
+  */ 
+
+#include QMK_KEYBOARD_H
+#include "joystick.h"
+#include "analog.h"
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+	[0] = LAYOUT(
+		KC_F1, KC_F2, KC_ESC, KC_1, KC_2, KC_3, KC_4, KC_5, 
+		KC_F3, KC_F4, KC_TAB, KC_Q, KC_W, KC_E, KC_R, KC_T, 
+		KC_F5, KC_F6, KC_NO, KC_A, KC_S, KC_D, KC_F, KC_G, 
+		KC_F7, KC_F8, KC_LSFT, KC_Z, KC_X, KC_C, KC_V, KC_B, 
+		KC_F9, KC_F10, LT(1, KC_NO), KC_LCTL, KC_LALT, LT(2, KC_NO), LT(1, KC_NO), KC_SPC),
+
+	[1] = LAYOUT(
+		RESET, DEBUG, KC_TILD, KC_6, KC_7, KC_8, KC_9, KC_0, 
+		KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, 
+		KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, 
+		KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, 
+		KC_F11, KC_F12, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS)
+};
+
+/* joystick config */
+joystick_config_t joystick_axes[JOYSTICK_AXES_COUNT] = {
+    [0] = JOYSTICK_AXIS_IN(F5, 1023, 512, 0),
+    [1] = JOYSTICK_AXIS_IN(F4, 0, 512, 1023)
+};
```

Comment:
One potential problem with placing this configuration in the keymap is that the configurator does not use the default keymap — it effectively generates a new `keymap.c` which contains only the `keymaps[]` array and nothing else. The firmware would still compile, because `quantum/joystick.c` supplies a weak definition of `joystick_axes[]`, but the joystick input won't work, because that weak definition does not specify any pins.

It could be possible to define `joystick_axes[]` at the keyboard level instead, but this will make replacing it at the keymap level difficult. I'm not really sure how to do this 100% properly (the presence of a weak definition in `quantum/joystick.c` prevents making the keyboard level definition weak, because the behavior of the linker with multiple weak definitions does not seem to be defined).

Reply:
I see.  Since the goal is to be able to use the configurator I moved the joystick functions up to the keyboard.  The initial implementation supports the 2 axis allowing the user to select their own thumbstick.  However in the future a user might want to swap the thumbstick for an encoder/mousewheel, linear throttles, or analog pressure sensitive buttons.  

At that point we'll have to make discrete keyboards or use the keymaps but loose configurator support.